Sprint 16. Project refactored to MVVM

ovseychik opened the pull request.

Commit:
Sprint 16. Initial commit

Commit:
Sprint 16. W-I-P commit

Commit:
Sprint 16. SettingsActivity refactored to MVVM

Commit:
Sprint 16. SearchActivity refactored to MVVM

Commit:
Sprint 16. SearchActivity refactored to MVVM

Commit:
Sprint 16. SearchActivity refactored to MVVM

Commit:
Sprint 16. PlayerActivity refactored to MVVM

ovseychik committed:
Create README.md

ovseychik committed:
Update README.md

Charmandik commented on app/src/main/java/com/example/playlistmaker/main/ui/activity/MainActivity.kt:
```diff
@@ -0,0 +1,48 @@
+import com.example.playlistmaker.search.ui.activity.SearchActivity
+
+class MainActivity : AppCompatActivity() {
+    private lateinit var binding: ActivityMainBinding
```
Рекомендация: Вместо использования lateinit, рекомендую использовать подход с добавлением nullability к типу объекта с оператором ? и сразу проинициализировать его значением null. Это решение позволяет избежать ошибок связанных с попытками доступа к неинициализированным объектам, которые могут привести к непредсказуемому поведению или крашам приложения. Хотя это может добавить некоторую дополнительную работу при обращении к объекту через оператор ?., но это значительно повысит безопасность приложения, так как вы будете уверены, что объект всегда либо имеет значение, либо равен null, и не будет вызывать исключительных ситуаций.

Можешь еще посмотреть на эту тему видео https://www.youtube.com/watch?v=0nXXUzMyF8c
или же прочитать статью https://habr.com/ru/companies/maxilect/articles/590061/

Charmandik commented on app/src/main/java/com/example/playlistmaker/player/ui/activity/PlayerActivity.kt:
```diff
@@ -0,0 +1,120 @@
+            trackName.text = track.trackName
+            artistName.text = track.artistName
+            time.text = millisFormat(track)
+            Glide.with(songArtwork)
```
отличный повод потренироваться писать функции расширения) ImageView.loadImage(url: String)

Charmandik commented on app/src/main/java/com/example/playlistmaker/player/ui/viewmodel/PlayerViewModel.kt:
```diff
@@ -0,0 +1,105 @@
+
+class PlayerViewModel(application: Application) : AndroidViewModel(application) {
+    companion object {
+        private const val REFRESH_TRACK_PROGRESS = 100L
```
При именовании констант для промежутка времени рекомендую всегда в конце добавлять единицу времени, например _MILLIS, _IN_SECONDS, _HOURS и т.д.

Charmandik commented on app/src/main/java/com/example/playlistmaker/player/ui/viewmodel/PlayerViewModel.kt:
```diff
@@ -0,0 +1,105 @@
+
+    private var statePlayerLiveData = MutableLiveData(PlayerState.STATE_PREPARED)
+
+    fun getStatePlayerLiveData(): LiveData<PlayerState> = statePlayerLiveData
```
Для ситуаций, когда нам нужно иметь два состояния( мутабельное и иммутабельное), рекомендую использовать подход с добавлением префикса в виде нижнего подчеркивания для мутабельного состояния(которое должно быть приватным):
private val _state = MutableLiveData<MyState>()
val state : LiveData<MyState> = _state

Так ты сможешь не придумывать разные имена, чтобы не повторяться, но и при этом нормально передавать смысл и сможешь точно понимать, что эта вот переменная это просто копия той, но иммутабельна

Charmandik commented on app/src/main/java/com/example/playlistmaker/search/data/network/RetrofitNetworkClient.kt:
```diff
@@ -0,0 +1,56 @@
+import retrofit2.converter.gson.GsonConverterFactory
+
+class RetrofitNetworkClient(private val context: Context) : NetworkClient {
+    private val iTunesBaseUrl = "https://itunes.apple.com"
```
это константа и лучше её закинуть в companion object

Commit:
Sprint 16.
- currentTimerLiveData merged into statePlayerLiveData.
- Repository interfaces moved to domain layer

Commit:
Merge remote-tracking branch 'origin/dev-16' into dev-16

Commit:
Sprint 16.
- currentTimerLiveData merged into statePlayerLiveData.
- Repository interfaces moved to domain layer

Commit:
Sprint 16.
- historyLiveData merged into searchStateLiveData

ovseychik commented:
* currentTimerLiveData объединена с statePlayerLiveData
* historyLiveData объединена с searchStateLiveData
* Перенес интерфейсы репозиториев в слой domain

Charmandik commented:
![image](https://github.com/ovseychik/PlaylistMaker/assets/34946350/7419703a-6335-4a6e-b642-8a6fda584934)
эти классы также нужно распределить по пакетам
